### PR TITLE
Optimize CUDA GPU

### PR DESCRIPTION
**Pull request by jean-m-cyr**
- etheminer does not support any CUDA architecture
  that requires the obsolete shuffle without sync.

- Copying mix to shared mem is expensive, do it
  only for actual solutions.

- Include whole work package in solution. Works out
  to about the same as including all of its components
  individually. Shortens construction of solution.

- redefine, resize, and realign results buffer such that
  cuda can use shift instead of multiply. As well as
  sequential addresses for mix copy.

- Delete unused cuda files, previously needed for
  pre-shuffle architectures.

### CODE REVIEW

File: libethash-cuda/dagger_shuffled.cuh [hunk 1]
```diff
@@ -114,11 +92,15 @@ __device__ __forceinline__ uint64_t compute_hash(
 			}
 		}
 	}
+
```
Comment (jean-m-cyr) [inline libethash-cuda/dagger_shuffled.cuh]: This is the actual optimization. Above is just cleanup

File: libethash-cuda/dagger_shuffled.cuh [hunk 2]
```diff
@@ -86,16 +74,6 @@ __device__ __forceinline__ uint64_t compute_hash(
 			uint32_t thread_mix = fnv_reduce(mix[p]);
 
 			// update mix accross threads
-#if CUDA_VERSION < SHUFFLE_DEPRECATED
```
Comment (chfast) [inline libethash-cuda/dagger_shuffled.cuh]: This is no longer needed?

Comment (jean-m-cyr) [reply libethash-cuda/dagger_shuffled.cuh]: Did I miss one?

**Comment by jean-m-cyr** [comment]
Also deleted 2 unused CUDA files

**Comment by jean-m-cyr** [comment]
@chfast Should I squash all of this? It might actually be easier to inspect one commit at a time?

**Comment by jean-m-cyr** [comment]
I wonder if we still need to support sm 30 & 35 Kepler architecture? Other than the Tesla K40 & K80 those were all 2GB or less GPUs. An I think the K40/80 have long since become unprofitable.

**Comment by chfast** [comment]
Isn't the K40 & K80 the ones some Macs have?

**Comment by chfast** [comment]
This is quite a lot of changes, let's better keep all the commits.

**Comment by MariusVanDerWijden** [comment]
I think we should still support 3.0 and 3.5 since those are still pretty common in big datacenters

File: libethcore/Farm.h [hunk 3]
```diff
@@ -338,10 +338,10 @@ class Farm: public FarmFace
 	 * @param _wp The WorkPackage that the Solution is for.
 	 * @return true iff the solution was good (implying that mining should be .
```
Comment (AaronOpfer) [inline libethcore/Farm.h]: This comment is no longer accurate.

Comment (jean-m-cyr) [reply libethcore/Farm.h]: Ah yes, thank you.

**Pull request merged**